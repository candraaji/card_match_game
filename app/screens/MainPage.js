import React, {useEffect} from 'react';
import {
  View,
  StyleSheet,
  Dimensions,
  ActivityIndicator,
  Text,
} from 'react-native';
import {composeCardGenerator} from '../utils/randomNumberHelper';
import Header from '../component/Header';
import CardList from '../component/Card/CardList';

const MainPage = () => {
  const [randomNumber, setRandomNumber] = React.useState([]);
  const [pickCard, setPickCard] = React.useState({});
  const [stepToOpen, setStepToOpen] = React.useState(0);
  const [countOpenCard, setCountOpenCard] = React.useState(1);
  const [loading, setLoading] = React.useState(false);
  const [isCompleted, setIsCompleted] = React.useState(false);

  useEffect(() => {
    (async () => {
      const dataCard = await composeCardGenerator();
      await setRandomNumber(dataCard);
    })();
  }, []);

  const onPressCard = card => {
    setCountOpenCard(x => x + 1);
    if (pickCard && pickCard?.index === card?.index) {
      setPickCard({});
    }
    setPickCard(card);

    if (
      pickCard &&
      pickCard?.number === card?.number &&
      pickCard?.index !== card?.index
    ) {
      const copyRandomNumber = randomNumber;
      copyRandomNumber.map(val => {
        if (val === card) {
          val.status = 'uneditable';
          val.position = 'show';
        }
        if (val === pickCard) {
          val.status = 'uneditable';
          val.position = 'show';
        }
      });
      setRandomNumber(copyRandomNumber);
    }

    if (countOpenCard >= 2) {
      setTimeout(() => {
        setCountOpenCard(1);
      }, 2000);
    }

    const detectCompleted = randomNumber.filter(
      val => val.status !== 'uneditable',
    );
    if (detectCompleted.length === 0) {
      setIsCompleted(true);
      setTimeout(() => {
        setIsCompleted(false);
        onPresRestart();
      }, 2000);
    }
  };

  const onPresRestart = async () => {
    await setLoading(true);
    await setPickCard({});
    await setStepToOpen(0);
    await setCountOpenCard(1);
    const dataCard = await composeCardGenerator();
    await setRandomNumber(dataCard);
    setTimeout(() => {
      setLoading(false);
    }, 500);
  };

  return (
    <>
      <Header stepToOpen={stepToOpen} onPresRestart={onPresRestart} />
      {isCompleted ? (
        <View style={styles.containerOthers}>
          <Text style={styles.textCongrations}>
            CONGRATULATIONS !! {'\n'} YOUR COMPLETED WITH {stepToOpen} STEP !
          </Text>
        </View>
      ) : loading ? (
        <View style={styles.containerOthers}>
          <ActivityIndicator size="large" />
        </View>
      ) : (
        <View style={styles.cardContainer}>
          <CardList
            listCard={randomNumber}
            setStepToOpen={setStepToOpen}
            onPressCard={onPressCard}
            countOpenCard={countOpenCard}
          />
        </View>
      )}
    </>
  );
};

const styles = StyleSheet.create({
  cardContainer: {
    height: Dimensions.get('window').height - 120,
    padding: '1%',
    justifyContent: 'space-around',
    flexWrap: 'wrap',
    alignItems: 'center',
    flexDirection: 'row',
  },
  containerOthers: {
    height: Dimensions.get('window').height - 120,
    justifyContent: 'center',
    flexWrap: 'wrap',
    alignContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  textCongrations: {
    color: 'white',
    textAlign: 'center',
  },
});

export default MainPage;
