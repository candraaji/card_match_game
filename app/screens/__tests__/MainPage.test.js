import React from 'react';
import renderer from 'react-test-renderer';
import {cleanup, act} from '@testing-library/react-native';

import MainPage from '../MainPage';

describe('MainPage Screens', () => {
  const dataCard = [
    {index: 1, number: 11, status: 'uneditable', position: 'show'},
    {index: 2, number: 11, status: 'uneditable', position: 'show'},
  ];
  const pickCard = {
    index: 1,
    number: 11,
    status: 'uneditable',
    position: 'show',
  };
  jest.useFakeTimers();

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  test('it should render MainPage screens correctly', () => {
    const wrapper = renderer.create(<MainPage />);
    act(() => {
      jest.runAllTimers();
    });
    expect(wrapper).toMatchSnapshot();
  });
  test('it should render MainPage Completed Screens', () => {
    React.useState = jest
      .fn()
      .mockReturnValueOnce([dataCard, {}])
      .mockReturnValueOnce([pickCard, {}])
      .mockReturnValueOnce([0, {}])
      .mockReturnValueOnce([1, {}])
      .mockReturnValueOnce([false, {}])
      .mockReturnValueOnce([true, {}]);

    const wrapper = renderer.create(<MainPage />);
    act(() => {
      jest.runAllTimers();
    });
    expect(wrapper).toMatchSnapshot();
  });
});
