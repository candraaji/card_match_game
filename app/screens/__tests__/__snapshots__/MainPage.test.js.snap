// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage Screens it should render MainPage Completed Screens 1`] = `
Array [
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": "5%",
        "minHeight": 50,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="restartButton"
    >
      <Text
        style={
          Object {
            "color": "white",
          }
        }
      >
        RESTART
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      STEPS : 
      0
    </Text>
  </View>,
  <View
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "height": 1214,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "textAlign": "center",
        }
      }
    >
      CONGRATULATIONS !! 
      

       YOUR COMPLETED WITH 
      0
       STEP !
    </Text>
  </View>,
]
`;

exports[`MainPage Screens it should render MainPage screens correctly 1`] = `
Array [
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": "5%",
        "minHeight": 50,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="restartButton"
    >
      <Text
        style={
          Object {
            "color": "white",
          }
        }
      >
        RESTART
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      STEPS : 
      0
    </Text>
  </View>,
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "height": 1214,
        "justifyContent": "space-around",
        "padding": "1%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="PressableCard"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 175,
            "justifyContent": "center",
            "padding": 10,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": 125,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/cardBackground.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 175,
              "justifyContent": "center",
              "width": 125,
            }
          }
        />
      </View>
    </View>
  </View>,
]
`;
